TypeSystem/variant.h

```diff
@@ -43,20 +52,91 @@ namespace current {
 
 struct BypassVariantTypeCheck {};
 
+template <typename NAME, typename TYPELIST>
+struct CurrentVariantNameHelper : CurrentVariant {
+  static std::string VariantName() { return NAME::VariantNameImpl(); }
+};
+
+template <typename TYPELIST>
+struct CurrentVariantNameHelper<reflection::CurrentVariantDefaultName, TYPELIST> : CurrentVariant {
+  static std::string VariantName() {
+    return reflection::CurrentVariantDefaultName::template VariantNameImpl<TYPELIST>();
+  }
+};
+
+namespace variant {
+
+#ifdef FEWER_COMPILE_TIME_CHECKS
+template <typename... TS>
+struct PopulateForAllTypes;
+template <>
+struct PopulateForAllTypes<> {
+  static void DoIt(std::unordered_map<std::type_index, const char*>&) {}
+};
+template <typename T, typename... TS>
+struct PopulateForAllTypes<T, TS...> {
+  static void DoIt(std::unordered_map<std::type_index, const char*>& result) {
+    const std::type_index ti = typeid(T);
+    CURRENT_ASSERT(!result.count(ti));
+    result[ti] = reflection::CurrentTypeNameAsConstCharPtr<T>();
+    PopulateForAllTypes<TS...>::DoIt(result);
+  }
+};
+template <typename... TS>
+struct RuntimeTypeListHelpersImpl {
+  using map_t = std::unordered_map<std::type_index, const char*>;
+  const map_t types_;
+  static map_t PopulateTypes() {
+    map_t result;
+    PopulateForAllTypes<TS...>::DoIt(result);
+    return result;
+  }
+  RuntimeTypeListHelpersImpl() : types_(PopulateTypes()) {}
+  template <typename T>
+  void AssertContains() const {
+    if (types_.find(std::type_index(typeid(T))) == types_.end()) {
+      std::cerr << "Variant type mismatch.\n\tType: " << reflection::CurrentTypeNameAsConstCharPtr<T>()
+                << "\nIs not among:";
+      for (const auto& type : types_) {
+        std::cerr << ' ' << type.second;
+      }
+      std::cerr << '\n';
```

Comment:
I thought of it -- the only downside it extra output of the test, and the upside is simplified debugging.

Perhaps `#ifndef NDEBUG` ?